### ISSUE TITLE
# Cuda Problem in Docker# ImportError: libcudart.so.10.0: cannot open shared object file: No such file or directory

### ISSUE THREAD

**Issue by QiangZiBro** [zju3dv/pvnet#118]
Dear author, I am a starter about docker, I hope you can help me to get the project run on my computer!
In order to make the environment deployed easier, I tried docker. But I still got some problems, here is what I did and found. 
## 1. rearrange the project directory as below:
```
.
├── ceres-solver-1.14.0.tar.gz
├── Dockerfile
├── how-to-docker.md
├── launch-docker.sh
├── pvnet
└── pvnet.tar.gz
```
## 2. Tried the Dockerfile in the project, some commands will cause permission error.
I guess the dockefile creates a user in the Linux OS, but some commands in the dockerfile acquire root. I modified some, but still can't solve the permission problem. So the dockerfile can't help me build an image.
**Error message:**
```
Step 19/22 : RUN source activate pvnet     && cd /home/$username/pvnet/lib/ransac_voting_gpu_layer     && python setup.py build_ext --inplace
 ---> Running in 50d11ee54f57
No CUDA runtime is found, using CUDA_HOME='/usr/local/cuda'
running build_ext
building 'ransac_voting' extension
creating build
error: could not create 'build': Permission denied
The command '/bin/bash -c source activate pvnet     && cd /home/$username/pvnet/lib/ransac_voting_gpu_layer     && python setup.py build_ext --inplace' returned a non-zero code: 1
```

**Modified dockerfile based on the project**
```
FROM nvidia/cuda:10.2-devel-ubuntu16.04
# NOTE: build this image from dir above Dockerfile location
# USE BASH
SHELL ["/bin/bash", "-c"]

## modify below
ARG username=pvnet
ARG groupid=1000
ARG userid=1000
## end modify

RUN apt-get update && apt-get install -y --no-install-recommends \
         build-essential \
         cmake \
         git \
         curl \
         wget \
         vim \
         ca-certificates \
         python-qt4 \
         libjpeg-dev \
         zip \
         unzip \
         sudo \
         libpng-dev \
         libeigen3-dev \
         libgoogle-glog-dev=0.3.4-0.1 \
         libsuitesparse-dev=1:4.4.6-1 \
         libatlas-base-dev=3.10.2-9 &&\
     rm -rf /var/lib/apt/lists/*

ENV LD_LIBRARY_PATH=/usr/local/cuda/lib64
ENV PYTHON_VERSION=3.6.7

RUN curl -o ~/miniconda.sh -O  https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh  && \
     chmod +x ~/miniconda.sh && \
     ~/miniconda.sh -b -p /opt/conda && \
     rm ~/miniconda.sh && \
    /opt/conda/bin/conda install conda-build

ENV PATH=$PATH:/opt/conda/bin/

RUN adduser --disabled-password --gecos '' $username \
    && adduser $username sudo \
    && echo '%sudo ALL=(ALL) NOPASSWD:ALL' >> /etc/sudoers
USER $username

# clone the repo and build ceres from source
ADD ./pvnet.tar.gz /home/$username/
COPY ./ceres-solver-1.14.0.tar.gz /home/$username/

RUN cd /home/$username \
    && cd pvnet \
    && sudo chmod a+rwx build_ceres_for_docker.sh \
    && sudo ./build_ceres_for_docker.sh /home/$username
ADD ./.condarc /home/$username/.condarc
ADD ./.pip /home/$username/.pip

# Create Enviroment
RUN cd /home/$username/pvnet \
    && sudo /opt/conda/bin/conda env create -f environment.yml

# follow remainder of steps in README
RUN source activate pvnet \
    && cd /home/$username/pvnet/lib/ransac_voting_gpu_layer \
    && python setup.py build_ext --inplace
RUN source activate pvnet \
    && cd /home/$username/pvnet/lib/utils/extend_utils \
    && python build_extend_utils_cffi_for_docker.py
ENV LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/$username/pvnet/lib/utils/extend_utils/lib

WORKDIR /home/$username

```

## 3. I wrote my own Dockerfile according to that of project, use root
The dockerfile is below, and I got maybe cuda version problem
* build commad `docker build --network=host -t pvnet-pytorch-cuda1002 .`
* run commad `docker run -it -e NVIDIA_VISIBLE_DEVICES=1 -v /home/qiangzibro/PVNET/pvnet:/workspace pvnet-pytorch-cuda1002` 

**the error message:**
```
(pvnet) root@0403d5f946b3:/pvnet# python tools/demo.py
Traceback (most recent call last):
  File "tools/demo.py", line 8, in <module>
    from lib.ransac_voting_gpu_layer.ransac_voting_gpu import ransac_voting_layer_v3
  File "/pvnet/lib/ransac_voting_gpu_layer/ransac_voting_gpu.py", line 2, in <module>
    import lib.ransac_voting_gpu_layer.ransac_voting as ransac_voting
ImportError: libcudart.so.10.0: cannot open shared object file: No such file or directory
```

**my dockerfile:**
```
FROM nvidia/cuda:10.2-devel-ubuntu16.04
# Use Cuda10.2
SHELL ["/bin/bash", "-c"]
RUN apt-get update && apt-get install -y --no-install-recommends \
         build-essential \
         cmake \
         git \
         curl \
         wget \
         vim \
         ca-certificates \
         python-qt4 \
         libjpeg-dev \
         zip \
         unzip \
         sudo \
         libpng-dev \
         libeigen3-dev \
         libgoogle-glog-dev=0.3.4-0.1 \
         libsuitesparse-dev=1:4.4.6-1 \
         privoxy \
         libatlas-base-dev=3.10.2-9 &&\
     rm -rf /var/lib/apt/lists/*

ENV LD_LIBRARY_PATH=/usr/local/cuda/lib64
ENV PYTHON_VERSION=3.6.7
ENV LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/pvnet/lib/utils/extend_utils/lib


# use conda to manage environment
COPY pvnet/miniconda.sh /root/
RUN  cd /root && \
     chmod +x miniconda.sh && \
     ./miniconda.sh -b -p /opt/conda && \
    /opt/conda/bin/conda install conda-build
ENV PATH=$PATH:/opt/conda/bin/

# add China source so that things can be downloaded faster
ADD pvnet/configs/.condarc /root/.condarc
ADD pvnet /pvnet/
ADD pvnet/configs/.pip /root/.pip

RUN cd /pvnet && ./build_ceres_for_docker.sh \
# 1.Set up python 3.6.7 environment
RUN cd /pvnet && conda env create -f environment.yml
# 2.Compile the Ransac Voting Layer
RUN source activate pvnet &&\
    cd /pvnet/lib/ransac_voting_gpu_layer \
    && python setup.py build_ext --inplace \
    && cd /pvnet/lib/utils/extend_utils \
    && python build_extend_utils_cffi_for_docker.py

# Solve `ImportError: cannot import name 'PILLOW_VERSION'`
RUN sed -i 's/, PILLOW_VERSION/, __version__ as PILLOW_VERSION/' /opt/conda/envs/pvnet/lib/python3.6/site-packages/torchvision/transforms/functional.py

RUN echo "source activate pvnet" >> /root/.bashrc

WORKDIR /pvnet
```

**Comment by pengsida** [comment]
I am not familiar with docker.
Use https://github.com/zju3dv/clean-pvnet if you want to make the environment deployed easier.

**Comment by hugovk** [comment]
This has been fixed in the new [torchvision 0.5.0](https://github.com/pytorch/vision/releases/tag/v0.5.0).

Please upgrade: `pip install torchvision>=0.5.0"`

**Comment by cuge1995** [comment]
> This has been fixed in the new [torchvision 0.5.0](https://github.com/pytorch/vision/releases/tag/v0.5.0).
> 
> Please upgrade: `pip install torchvision>=0.5.0"`

Not solve my problem
